src/Microsoft.ML.Data/Transforms/ValueMapping.cs

```diff
@@ -396,29 +414,32 @@ internal sealed class Options
             : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(ValueMappingTransformer)), columns)
         {
-            Host.CheckNonEmpty(keyColumn, nameof(keyColumn), "A key column must be specified when passing in an IDataView for the value mapping");
-            Host.CheckNonEmpty(valueColumn, nameof(valueColumn), "A value column must be specified when passing in an IDataView for the value mapping");
-            _valueMap = CreateValueMapFromDataView(lookupMap, keyColumn, valueColumn);
-            int valueColumnIdx = 0;
-            Host.Check(lookupMap.Schema.TryGetColumnIndex(valueColumn, out valueColumnIdx));
-            ValueColumnMetadata = lookupMap.Schema[valueColumnIdx].Annotations;
+            _lookupKeyColumn = lookupKeyColumn;
+            _lookupValueColumn = lookupValueColumn;
+            _valueMap = CreateValueMapFromDataView(lookupMap, _lookupKeyColumn, _lookupValueColumn);
+            ValueColumnMetadata = lookupMap.Schema[_lookupValueColumn.Index].Annotations;
 
             // Create the byte array of the original IDataView, this is used for saving out the data.
-            _dataView = GetBytesFromDataView(Host, lookupMap, keyColumn, valueColumn);
+            _dataView = GetBytesFromDataView(Host, lookupMap, lookupKeyColumn.Name, lookupValueColumn.Name);
         }
 
-        private ValueMap CreateValueMapFromDataView(IDataView dataView, string keyColumn, string valueColumn)
+        private ValueMap CreateValueMapFromDataView(IDataView dataView, DataViewSchema.Column keyColumn, DataViewSchema.Column valueColumn)
         {
-            // Confirm that the key and value columns exist in the dataView
-            Host.Check(dataView.Schema.TryGetColumnIndex(keyColumn, out int keyIdx), "Key column " + keyColumn + " does not exist in the given dataview");
-            Host.Check(dataView.Schema.TryGetColumnIndex(valueColumn, out int valueIdx), "Value column " + valueColumn + " does not exist in the given dataview");
-            var keyType = dataView.Schema[keyIdx].Type;
-            var valueType = dataView.Schema[valueIdx].Type;
-            var valueMap = ValueMap.Create(keyType, valueType, ValueColumnMetadata);
-            using (var cursor = dataView.GetRowCursor(dataView.Schema[keyIdx], dataView.Schema[valueIdx]))
+            var retrievedKeyColumn = dataView.Schema[keyColumn.Index];
+            Host.Check(retrievedKeyColumn.Index == keyColumn.Index, nameof(keyColumn) + "'s column index doesn't match that of the associated column in " + nameof(dataView));
+            Host.Check(retrievedKeyColumn.Type == keyColumn.Type, nameof(keyColumn) + "'s column type doesn't match that of the associated column in " + nameof(dataView));
+            Host.Check(retrievedKeyColumn.Annotations == keyColumn.Annotations, nameof(keyColumn) + "'s column annotions don't match those of the associated column in " + nameof(dataView));
+
+            var retrievedValueColumn = dataView.Schema[valueColumn.Index];
```

Comment:
same here.

Reply:
Will add a check as well.

---
In reply to: [271535097](https://github.com/dotnet/machinelearning/pull/3180#discussion_r271535097) [](ancestors = 271535097)